src/main/scala/Clock.scala

```diff
@@ -2,273 +2,199 @@ package chess
   }
 
-  def stop = PausedClock(
-    config = config,
-    color = color,
-    whiteTime = color.fold(whiteTime + (timer to now), whiteTime),
-    blackTime = color.fold(blackTime, blackTime + (timer to now)),
-    whiteBerserk = whiteBerserk,
-    blackBerserk = blackBerserk
-  )
-
-  def addTime(c: Color, t: Centis): RunningClock = c match {
-    case White => copy(whiteTime = whiteTime + t)
-    case Black => copy(blackTime = blackTime + t)
+  def goBerserk(c: Color) = updatePlayer(c) { p =>
+    if (p.berserk) p
+    else p.copy(
+      berserk = true,
+      limit = p.limit - berserkPenalty
+    )
   }
 
-  def giveTime(c: Color, t: Centis): RunningClock = addTime(c, -t)
+  protected def berserkPenalty =
+    if (limitSeconds < 40 * incrementSeconds) Centis(0)
+    else Centis(limitSeconds * (100 / 2))
 
-  def berserk(c: Color): RunningClock = addTime(c, Clock.berserkPenalty(this, color)).copy(
-    whiteBerserk = whiteBerserk || c.white,
-    blackBerserk = blackBerserk || c.black
-  )
+  def incrementOf(c: Color) = if (berserked(c)) Centis(0) else increment
 
-  def switch: RunningClock = copy(color = !color)
+  def berserked(c: Color) = players(c).berserk
+  def lag(c: Color) = players(c).lag
 
-  def takeback: RunningClock = {
-    copy(
-      color = !color,
-      timer = now
-    )
-  }
+  def emergSeconds = config.emergSeconds
+  def estimateTotalSeconds = config.estimateTotalSeconds
+  def estimateTotalTime = config.estimateTotalTime
+  def increment = config.increment
+  def incrementSeconds = config.incrementSeconds
+  def limit = config.limit
+  def limitInMinutes = config.limitInMinutes
+  def limitSeconds = config.limitSeconds
 }
 
-case class PausedClock(
-    config: Clock.Config,
-    color: Color,
-    whiteTime: Centis,
-    blackTime: Centis,
-    whiteBerserk: Boolean,
-    blackBerserk: Boolean
-) extends Clock {
-
-  val timerOption = None
+case class ClockPlayer(
+    limit: Centis,
+    elapsed: Centis = Centis(0),
+    lag: Centis = Centis(0),
+    berserk: Boolean = false
+) {
+  import ClockPlayer._
 
-  def stop = this
+  def isInit = elapsed.centis == 0
 
-  def addTime(c: Color, t: Centis): PausedClock = c match {
-    case White => copy(whiteTime = whiteTime + t)
-    case Black => copy(blackTime = blackTime + t)
-  }
+  def adjust(elapsedDelta: Centis, limitDelta: Centis = Centis(0)) = copy(
+    elapsed = elapsed + elapsedDelta,
+    limit = limit + limitDelta
+  )
 
-  def giveTime(c: Color, t: Centis): PausedClock = addTime(c, -t)
+  def remaining = limit - elapsed
 
-  def berserk(c: Color): PausedClock = addTime(c, Clock.berserkPenalty(this, color)).copy(
-    whiteBerserk = c.fold(true, whiteBerserk),
-    blackBerserk = c.fold(blackBerserk, true)
-  )
+  def giveTime(t: Centis) = copy(limit = limit + t)
 
-  def switch: PausedClock = copy(color = !color)
+  def lagComp(lagOpt: Option[Centis]): Centis = lagOpt match {
+    case None => Centis(0)
+    case Some(lag) => lag atMost maxLagComp
+  }
```

Comment:
point of detail, using `Option.fold` is better when possible:

`lagOpt.fold(Centis(0))(_ atMost maxLagComp)`